File: snippets/screens/flutter/restaurants_list_screen/1/restaurants_list_screen.dart
import 'package:flutter/material.dart';

class RestaurantsListScreen extends StatefulWidget {
  const RestaurantsListScreen({super.key});

  @override
  State<RestaurantsListScreen> createState() => _RestaurantsListScreenState();
}

class _RestaurantsListScreenState extends State<RestaurantsListScreen> {
  final TextEditingController _searchController = TextEditingController();
  String _searchQuery = '';
  String _selectedFilter = 'All';
  String _selectedSort = 'Rating';

  final List<Restaurant> _restaurants = [
    Restaurant(
      id: '1',
      name: 'The Golden Spoon',
      cuisine: 'Italian',
      rating: 4.8,
      reviewCount: 324,
      priceRange: PriceRange.expensive,
      deliveryTime: '25-40 min',
      deliveryFee: 3.99,
      distance: 0.8,
      imageUrl: 'https://via.placeholder.com/300x200',
      isOpen: true,
      isFeatured: true,
      address: '123 Main St, Downtown',
      phoneNumber: '+1 555-0123',
    ),
    Restaurant(
      id: '2',
      name: 'Sakura Sushi',
      cuisine: 'Japanese',
      rating: 4.6,
      reviewCount: 198,
      priceRange: PriceRange.moderate,
      deliveryTime: '30-45 min',
      deliveryFee: 2.99,
      distance: 1.2,
      imageUrl: 'https://via.placeholder.com/300x200',
      isOpen: true,
      isFeatured: false,
      address: '456 Oak Ave, Midtown',
      phoneNumber: '+1 555-0124',
    ),
    Restaurant(
      id: '3',
      name: 'Taco Libre',
      cuisine: 'Mexican',
      rating: 4.4,
      reviewCount: 256,
      priceRange: PriceRange.budget,
      deliveryTime: '15-30 min',
      deliveryFee: 1.99,
      distance: 0.5,
      imageUrl: 'https://via.placeholder.com/300x200',
      isOpen: true,
      isFeatured: true,
      address: '789 Pine St, Eastside',
      phoneNumber: '+1 555-0125',
    ),
    Restaurant(
      id: '4',
      name: 'Le Petit Bistro',
      cuisine: 'French',
      rating: 4.9,
      reviewCount: 412,
      priceRange: PriceRange.expensive,
      deliveryTime: '35-50 min',
      deliveryFee: 4.99,
      distance: 2.1,
      imageUrl: 'https://via.placeholder.com/300x200',
      isOpen: false,
      isFeatured: false,
      address: '321 Elm Dr, Westside',
      phoneNumber: '+1 555-0126',
    ),
    Restaurant(
      id: '5',
      name: 'Spice Garden',
      cuisine: 'Indian',
      rating: 4.5,
      reviewCount: 189,
      priceRange: PriceRange.moderate,
      deliveryTime: '25-40 min',
      deliveryFee: 2.49,
      distance: 1.5,
      imageUrl: 'https://via.placeholder.com/300x200',
      isOpen: true,
      isFeatured: false,
      address: '654 Maple Rd, Northside',
      phoneNumber: '+1 555-0127',
    ),
    Restaurant(
      id: '6',
      name: 'Dragon Palace',
      cuisine: 'Chinese',
      rating: 4.3,
      reviewCount: 298,
      priceRange: PriceRange.budget,
      deliveryTime: '20-35 min',
      deliveryFee: 1.49,
      distance: 1.0,
      imageUrl: 'https://via.placeholder.com/300x200',
      isOpen: true,
      isFeatured: false,
      address: '987 Cedar St, Southside',
      phoneNumber: '+1 555-0128',
    ),
  ];

  List<Restaurant> get _filteredRestaurants {
    List<Restaurant> filtered = _restaurants;

    // Apply search filter
    if (_searchQuery.isNotEmpty) {
      filtered = filtered.where((restaurant) =>
          restaurant.name.toLowerCase().contains(_searchQuery.toLowerCase()) ||
          restaurant.cuisine.toLowerCase().contains(_searchQuery.toLowerCase())
      ).toList();
    }

    // Apply cuisine filter
    if (_selectedFilter != 'All') {
      filtered = filtered.where((restaurant) => restaurant.cuisine == _selectedFilter).toList();
    }

    // Apply sorting
    switch (_selectedSort) {
      case 'Rating':
        filtered.sort((a, b) => b.rating.compareTo(a.rating));
        break;
      case 'Distance':
        filtered.sort((a, b) => a.distance.compareTo(b.distance));
        break;
      case 'Delivery Time':
        filtered.sort((a, b) => a.deliveryTime.compareTo(b.deliveryTime));
        break;
      case 'Price':
        filtered.sort((a, b) => a.priceRange.index.compareTo(b.priceRange.index));
        break;
    }

    // Prioritize featured and open restaurants
    filtered.sort((a, b) {
      if (a.isFeatured && !b.isFeatured) return -1;
      if (!a.isFeatured && b.isFeatured) return 1;
      if (a.isOpen && !b.isOpen) return -1;
      if (!a.isOpen && b.isOpen) return 1;
      return 0;
    });

    return filtered;
  }

  @override
  Widget build(BuildContext context) {
    final cuisines = ['All', ...{'Italian', 'Japanese', 'Mexican', 'French', 'Indian', 'Chinese'}];

    return Scaffold(
      appBar: AppBar(
        title: const Text('Restaurants'),
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
        actions: [
          IconButton(
            icon: const Icon(Icons.map),
            onPressed: () => _showMapView(),
          ),
          IconButton(
            icon: const Icon(Icons.filter_list),
            onPressed: () => _showFilterDialog(),
          ),
        ],
      ),
      body: Column(
        children: [
          // Search Bar
          Padding(
            padding: const EdgeInsets.all(16.0),
            child: TextField(
              controller: _searchController,
              decoration: InputDecoration(
                hintText: 'Search restaurants or cuisine...',
                prefixIcon: const Icon(Icons.search),
                suffixIcon: _searchQuery.isNotEmpty
                    ? IconButton(
                        icon: const Icon(Icons.clear),
                        onPressed: () {
                          _searchController.clear();
                          setState(() {
                            _searchQuery = '';
                          });
                        },
                      )
                    : null,
                border: OutlineInputBorder(
                  borderRadius: BorderRadius.circular(12),
                ),
              ),
              onChanged: (value) {
                setState(() {
                  _searchQuery = value;
                });
              },
            ),
          ),

          // Filter Chips
          SizedBox(
            height: 50,
            child: ListView.builder(
              scrollDirection: Axis.horizontal,
              padding: const EdgeInsets.symmetric(horizontal: 16),
              itemCount: cuisines.length + 1, // +1 for sort chip
              itemBuilder: (context, index) {
                if (index == cuisines.length) {
                  return Padding(
                    padding: const EdgeInsets.only(left: 8),
                    child: ActionChip(
                      avatar: const Icon(Icons.sort, size: 18),
                      label: Text('Sort: $_selectedSort'),
                      onPressed: () => _showSortDialog(),
                    ),
                  );
                }

                final cuisine = cuisines[index];
                return Padding(
                  padding: const EdgeInsets.only(right: 8),
                  child: FilterChip(
                    label: Text(cuisine),
                    selected: _selectedFilter == cuisine,
                    onSelected: (selected) {
                      setState(() {
                        _selectedFilter = cuisine;
                      });
                    },
                    selectedColor: Colors.orange.withOpacity(0.2),
                  ),
                );
              },
            ),
          ),

          const SizedBox(height: 8),

          // Restaurants List
          Expanded(
            child: _filteredRestaurants.isEmpty
                ? _buildEmptyState()
                : ListView.builder(
                    padding: const EdgeInsets.symmetric(horizontal: 16),
                    itemCount: _filteredRestaurants.length,
                    itemBuilder: (context, index) {
                      final restaurant = _filteredRestaurants[index];
                      return _buildRestaurantCard(restaurant);
                    },
                  ),
          ),
        ],
      ),
    );
  }

  Widget _buildEmptyState() {
    return Center(
      child: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: [
          Icon(
            Icons.restaurant,
            size: 64,
            color: Colors.grey[400],
          ),
          const SizedBox(height: 16),
          Text(
            _searchQuery.isEmpty ? 'No restaurants available' : 'No restaurants found',
            style: Theme.of(context).textTheme.titleMedium?.copyWith(
                  color: Colors.grey[600],
                ),
          ),
          const SizedBox(height: 8),
          Text(
            _searchQuery.isEmpty
                ? 'Try changing your location or check back later'
                : 'Try adjusting your search or filters',
            style: TextStyle(color: Colors.grey[500]),
            textAlign: TextAlign.center,
          ),
        ],
      ),
    );
  }

  Widget _buildRestaurantCard(Restaurant restaurant) {
    return Card(
      margin: const EdgeInsets.only(bottom: 16),
      clipBehavior: Clip.antiAlias,
      child: InkWell(
        onTap: () => _navigateToRestaurantDetail(restaurant),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            // Restaurant Image
            Stack(
              children: [
                Container(
                  height: 200,
                  width: double.infinity,
                  decoration: BoxDecoration(
                    gradient: LinearGradient(
                      colors: [
                        _getCuisineColor(restaurant.cuisine).withOpacity(0.8),
                        _getCuisineColor(restaurant.cuisine),
                      ],
                      begin: Alignment.topLeft,
                      end: Alignment.bottomRight,
                    ),
                  ),
                  child: const Center(
                    child: Icon(
                      Icons.restaurant,
                      size: 48,
                      color: Colors.white,
                    ),
                  ),
                ),

                // Status badges
                Positioned(
                  top: 12,
                  left: 12,
                  child: Column(
                    crossAxisAlignment: CrossAxisAlignment.start,
                    children: [
                      if (restaurant.isFeatured)
                        Container(
                          padding: const EdgeInsets.symmetric(horizontal: 8, vertical: 4),
                          decoration: BoxDecoration(
                            color: Colors.orange,
                            borderRadius: BorderRadius.circular(12),
                          ),
                          child: const Text(
                            'Featured',
                            style: TextStyle(
                              color: Colors.white,
                              fontSize: 10,
                              fontWeight: FontWeight.bold,
                            ),
                          ),
                        ),
                      if (!restaurant.isOpen)
                        Container(
                          margin: EdgeInsets.only(top: restaurant.isFeatured ? 6 : 0),
                          padding: const EdgeInsets.symmetric(horizontal: 8, vertical: 4),
                          decoration: BoxDecoration(
                            color: Colors.red,
                            borderRadius: BorderRadius.circular(12),
                          ),
                          child: const Text(
                            'Closed',
                            style: TextStyle(
                              color: Colors.white,
                              fontSize: 10,
                              fontWeight: FontWeight.bold,
                            ),
                          ),
                        ),
                    ],
                  ),
                ),

                // Distance and delivery fee
                Positioned(
                  top: 12,
                  right: 12,
                  child: Container(
                    padding: const EdgeInsets.symmetric(horizontal: 8, vertical: 4),
                    decoration: BoxDecoration(
                      color: Colors.black.withOpacity(0.7),
                      borderRadius: BorderRadius.circular(12),
                    ),
                    child: Text(
                      '${restaurant.distance.toStringAsFixed(1)} km',
                      style: const TextStyle(
                        color: Colors.white,
                        fontSize: 12,
                        fontWeight: FontWeight.w600,
                      ),
                    ),
                  ),
                ),
              ],
            ),

            // Restaurant Info
            Padding(
              padding: const EdgeInsets.all(16),
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  Row(
                    children: [
                      Expanded(
                        child: Text(
                          restaurant.name,
                          style: const TextStyle(
                            fontSize: 18,
                            fontWeight: FontWeight.bold,
                          ),
                        ),
                      ),
                      Container(
                        padding: const EdgeInsets.symmetric(horizontal: 6, vertical: 2),
                        decoration: BoxDecoration(
                          color: _getPriceColor(restaurant.priceRange).withOpacity(0.1),
                          borderRadius: BorderRadius.circular(6),
                        ),
                        child: Text(
                          _getPriceRangeText(restaurant.priceRange),
                          style: TextStyle(
                            color: _getPriceColor(restaurant.priceRange),
                            fontWeight: FontWeight.bold,
                            fontSize: 12,
                          ),
                        ),
                      ),
                    ],
                  ),

                  const SizedBox(height: 4),

                  Text(
                    restaurant.cuisine,
                    style: TextStyle(
                      color: _getCuisineColor(restaurant.cuisine),
                      fontWeight: FontWeight.w600,
                      fontSize: 14,
                    ),
                  ),

                  const SizedBox(height: 8),

                  Row(
                    children: [
                      Icon(
                        Icons.star,
                        color: Colors.amber,
                        size: 16,
                      ),
                      const SizedBox(width: 4),
                      Text(
                        restaurant.rating.toString(),
                        style: const TextStyle(
                          fontWeight: FontWeight.w600,
                        ),
                      ),
                      Text(
                        ' (${restaurant.reviewCount})',
                        style: TextStyle(
                          color: Colors.grey[600],
                        ),
                      ),

                      const Spacer(),

                      Icon(
                        Icons.access_time,
                        color: Colors.grey[600],
                        size: 16,
                      ),
                      const SizedBox(width: 4),
                      Text(
                        restaurant.deliveryTime,
                        style: TextStyle(
                          color: Colors.grey[600],
                          fontWeight: FontWeight.w500,
                        ),
                      ),
                    ],
                  ),

                  const SizedBox(height: 8),

                  Row(
                    children: [
                      Icon(
                        Icons.location_on,
                        color: Colors.grey[600],
                        size: 16,
                      ),
                      const SizedBox(width: 4),
                      Expanded(
                        child: Text(
                          restaurant.address,
                          style: TextStyle(
                            color: Colors.grey[600],
                            fontSize: 12,
                          ),
                          maxLines: 1,
                          overflow: TextOverflow.ellipsis,
                        ),
                      ),

                      const SizedBox(width: 8),

                      Text(
                        'Delivery: \$${restaurant.deliveryFee.toStringAsFixed(2)}',
                        style: TextStyle(
                          color: Colors.green[700],
                          fontWeight: FontWeight.w600,
                          fontSize: 12,
                        ),
                      ),
                    ],
                  ),

                  const SizedBox(height: 12),

                  // Action Buttons
                  Row(
                    children: [
                      Expanded(
                        child: OutlinedButton.icon(
                          onPressed: restaurant.isOpen ? () => _addToFavorites(restaurant) : null,
                          icon: const Icon(Icons.favorite_border, size: 16),
                          label: const Text('Save'),
                          style: OutlinedButton.styleFrom(
                            foregroundColor: Colors.red,
                          ),
                        ),
                      ),
                      const SizedBox(width: 8),
                      Expanded(
                        flex: 2,
                        child: ElevatedButton.icon(
                          onPressed: restaurant.isOpen
                              ? () => _navigateToRestaurantDetail(restaurant)
                              : null,
                          icon: const Icon(Icons.restaurant_menu, size: 16),
                          label: Text(restaurant.isOpen ? 'View Menu' : 'Closed'),
                          style: ElevatedButton.styleFrom(
                            backgroundColor: restaurant.isOpen
                                ? _getCuisineColor(restaurant.cuisine)
                                : Colors.grey,
                            foregroundColor: Colors.white,
                          ),
                        ),
                      ),
                    ],
                  ),
                ],
              ),
            ),
          ],
        ),
      ),
    );
  }

  Color _getCuisineColor(String cuisine) {
    switch (cuisine) {
      case 'Italian':
        return Colors.red;
      case 'Japanese':
        return Colors.pink;
      case 'Mexican':
        return Colors.orange;
      case 'French':
        return Colors.purple;
      case 'Indian':
        return Colors.deepOrange;
      case 'Chinese':
        return Colors.red.shade800;
      default:
        return Colors.blue;
    }
  }

  Color _getPriceColor(PriceRange range) {
    switch (range) {
      case PriceRange.budget:
        return Colors.green;
      case PriceRange.moderate:
        return Colors.orange;
      case PriceRange.expensive:
        return Colors.red;
    }
  }

  String _getPriceRangeText(PriceRange range) {
    switch (range) {
      case PriceRange.budget:
        return '\$';
      case PriceRange.moderate:
        return '\$\$';
      case PriceRange.expensive:
        return '\$\$\$';
    }
  }

  void _navigateToRestaurantDetail(Restaurant restaurant) {
    Navigator.of(context).push(
      MaterialPageRoute(
        builder: (context) => RestaurantDetailScreen(restaurant: restaurant),
      ),
    );
  }

  void _addToFavorites(Restaurant restaurant) {
    ScaffoldMessenger.of(context).showSnackBar(
      SnackBar(content: Text('${restaurant.name} added to favorites!')),
    );
  }

  void _showMapView() {
    ScaffoldMessenger.of(context).showSnackBar(
      const SnackBar(content: Text('Map view would be implemented here')),
    );
  }

  void _showFilterDialog() {
    showModalBottomSheet(
      context: context,
      builder: (context) => Container(
        padding: const EdgeInsets.all(20),
        child: Column(
          mainAxisSize: MainAxisSize.min,
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            const Text(
              'Filter Options',
              style: TextStyle(fontSize: 18, fontWeight: FontWeight.bold),
            ),
            const SizedBox(height: 16),
            const Text('Price Range:'),
            const SizedBox(height: 8),
            Wrap(
              spacing: 8,
              children: PriceRange.values.map((range) {
                return FilterChip(
                  label: Text(_getPriceRangeText(range)),
                  selected: false,
                  onSelected: (selected) {
                    // Apply price filter
                  },
                );
              }).toList(),
            ),
            const SizedBox(height: 16),
            const Text('Rating:'),
            const SizedBox(height: 8),
            Wrap(
              spacing: 8,
              children: ['4.0+', '4.5+', '4.8+'].map((rating) {
                return FilterChip(
                  label: Text(rating),
                  selected: false,
                  onSelected: (selected) {
                    // Apply rating filter
                  },
                );
              }).toList(),
            ),
          ],
        ),
      ),
    );
  }

  void _showSortDialog() {
    showDialog(
      context: context,
      builder: (context) => AlertDialog(
        title: const Text('Sort By'),
        content: Column(
          mainAxisSize: MainAxisSize.min,
          children: ['Rating', 'Distance', 'Delivery Time', 'Price'].map((option) {
            return RadioListTile<String>(
              title: Text(option),
              value: option,
              groupValue: _selectedSort,
              onChanged: (value) {
                setState(() {
                  _selectedSort = value!;
                });
                Navigator.of(context).pop();
              },
            );
          }).toList(),
        ),
      ),
    );
  }

  @override
  void dispose() {
    _searchController.dispose();
    super.dispose();
  }
}

enum PriceRange { budget, moderate, expensive }

class Restaurant {
  final String id;
  final String name;
  final String cuisine;
  final double rating;
  final int reviewCount;
  final PriceRange priceRange;
  final String deliveryTime;
  final double deliveryFee;
  final double distance;
  final String imageUrl;
  final bool isOpen;
  final bool isFeatured;
  final String address;
  final String phoneNumber;

  Restaurant({
    required this.id,
    required this.name,
    required this.cuisine,
    required this.rating,
    required this.reviewCount,
    required this.priceRange,
    required this.deliveryTime,
    required this.deliveryFee,
    required this.distance,
    required this.imageUrl,
    required this.isOpen,
    required this.isFeatured,
    required this.address,
    required this.phoneNumber,
  });
}

// This would typically be in a separate file
class RestaurantDetailScreen extends StatelessWidget {
  final Restaurant restaurant;

  const RestaurantDetailScreen({super.key, required this.restaurant});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(restaurant.name),
      ),
      body: const Center(
        child: Text('Restaurant Detail Screen'),
      ),
    );
  }
}

File: snippets/screens/swiftui/restaurants_list_screen/3/restaurants_list_screen.swift
import SwiftUI

struct RestaurantsListScreen: View {
    @State private var searchText = ""
    @State private var selectedSort = "Rating"
    @State private var showingFilter = false

    let restaurants = [
        Restaurant(id: "1", name: "The Golden Spoon", cuisine: "Italian", rating: 4.8, reviewCount: 324, priceRange: .expensive, deliveryTime: "25-40 min", deliveryFee: 3.99, distance: 0.8, isOpen: true, isFeatured: true, address: "123 Main St, Downtown"),
        Restaurant(id: "2", name: "Sakura Sushi", cuisine: "Japanese", rating: 4.6, reviewCount: 198, priceRange: .moderate, deliveryTime: "30-45 min", deliveryFee: 2.99, distance: 1.2, isOpen: true, isFeatured: false, address: "456 Oak Ave, Midtown"),
        Restaurant(id: "3", name: "Taco Libre", cuisine: "Mexican", rating: 4.4, reviewCount: 256, priceRange: .budget, deliveryTime: "15-30 min", deliveryFee: 1.99, distance: 0.5, isOpen: true, isFeatured: true, address: "789 Pine St, Eastside"),
        Restaurant(id: "4", name: "Le Petit Bistro", cuisine: "French", rating: 4.9, reviewCount: 412, priceRange: .expensive, deliveryTime: "35-50 min", deliveryFee: 4.99, distance: 2.1, isOpen: false, isFeatured: false, address: "321 Elm Dr, Westside")
    ]

    var filteredRestaurants: [Restaurant] {
        var filtered = restaurants

        if !searchText.isEmpty {
            filtered = filtered.filter { restaurant in
                restaurant.name.localizedCaseInsensitiveContains(searchText) ||
                restaurant.cuisine.localizedCaseInsensitiveContains(searchText)
            }
        }

        filtered = filtered.sorted { first, second in
            if first.isFeatured && !second.isFeatured { return true }
            if !first.isFeatured && second.isFeatured { return false }
            if first.isOpen && !second.isOpen { return true }
            if !first.isOpen && second.isOpen { return false }

            switch selectedSort {
            case "Rating": return first.rating > second.rating
            case "Distance": return first.distance < second.distance
            case "Price": return first.priceRange.rawValue < second.priceRange.rawValue
            default: return first.rating > second.rating
            }
        }

        return filtered
    }

    var body: some View {
        NavigationStack {
            VStack(spacing: 0) {
                // Search Bar
                SearchBar(text: $searchText)
                    .padding(.horizontal)
                    .padding(.top, 8)

                // Filter Chips
                ScrollView(.horizontal, showsIndicators: false) {
                    HStack(spacing: 12) {
                        SortChip(selectedSort: $selectedSort, title: "Sort: \(selectedSort)")

                        FilterChip(title: "Italian", isSelected: false) { }
                        FilterChip(title: "Fast Food", isSelected: false) { }
                        FilterChip(title: "Open Now", isSelected: false) { }
                    }
                    .padding(.horizontal)
                }
                .padding(.vertical, 8)

                if filteredRestaurants.isEmpty {
                    // Empty State
                    VStack(spacing: 16) {
                        Spacer()
                        Image(systemName: "fork.knife.circle")
                            .font(.system(size: 64))
                            .foregroundColor(.gray.opacity(0.5))
                        Text("No restaurants found")
                            .font(.title2)
                            .foregroundColor(.secondary)
                        Text("Try adjusting your search or filters")
                            .font(.caption)
                            .foregroundColor(.secondary)
                        Spacer()
                    }
                } else {
                    // Restaurants List
                    ScrollView {
                        LazyVStack(spacing: 16) {
                            ForEach(filteredRestaurants) { restaurant in
                                NavigationLink(destination: RestaurantDetailScreen(restaurant: restaurant)) {
                                    RestaurantCard(restaurant: restaurant)
                                }
                                .buttonStyle(PlainButtonStyle())
                            }
                        }
                        .padding(.horizontal)
                        .padding(.top, 8)
                    }
                }
            }
            .navigationTitle("Restaurants")
            .toolbar {
                ToolbarItemGroup(placement: .navigationBarTrailing) {
                    Button {
                        // Show map view
                    } label: {
                        Image(systemName: "map")
                    }

                    Button {
                        showingFilter = true
                    } label: {
                        Image(systemName: "line.3.horizontal.decrease.circle")
                    }
                }
            }
            .sheet(isPresented: $showingFilter) {
                FilterSheet()
            }
        }
    }
}

struct RestaurantCard: View {
    let restaurant: Restaurant

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            // Restaurant Header
            ZStack(alignment: .topLeading) {
                RoundedRectangle(cornerRadius: 12)
                    .fill(restaurant.cuisineColor.gradient)
                    .frame(height: 200)

                // Badges
                VStack(alignment: .leading, spacing: 6) {
                    if restaurant.isFeatured {
                        Text("Featured")
                            .font(.caption2)
                            .fontWeight(.bold)
                            .padding(.horizontal, 8)
                            .padding(.vertical, 4)
                            .background(Color.orange)
                            .foregroundColor(.white)
                            .clipShape(RoundedRectangle(cornerRadius: 8))
                    }

                    if !restaurant.isOpen {
                        Text("Closed")
                            .font(.caption2)
                            .fontWeight(.bold)
                            .padding(.horizontal, 8)
                            .padding(.vertical, 4)
                            .background(Color.red)
                            .foregroundColor(.white)
                            .clipShape(RoundedRectangle(cornerRadius: 8))
                    }
                }
                .padding(12)

                // Distance badge (top right)
                HStack {
                    Spacer()
                    Text("\(restaurant.distance, specifier: "%.1f") km")
                        .font(.caption)
                        .fontWeight(.semibold)
                        .padding(.horizontal, 8)
                        .padding(.vertical, 4)
                        .background(Color.black.opacity(0.7))
                        .foregroundColor(.white)
                        .clipShape(RoundedRectangle(cornerRadius: 8))
                        .padding(12)
                }
            }

            // Restaurant Info
            VStack(alignment: .leading, spacing: 12) {
                HStack {
                    Text(restaurant.name)
                        .font(.headline)
                        .fontWeight(.bold)

                    Spacer()

                    Text(restaurant.priceRange.displayText)
                        .font(.caption)
                        .fontWeight(.bold)
                        .padding(.horizontal, 6)
                        .padding(.vertical, 2)
                        .background(restaurant.priceRange.color.opacity(0.1))
                        .foregroundColor(restaurant.priceRange.color)
                        .clipShape(RoundedRectangle(cornerRadius: 4))
                }

                Text(restaurant.cuisine)
                    .font(.subheadline)
                    .foregroundColor(restaurant.cuisineColor)
                    .fontWeight(.semibold)

                HStack {
                    // Rating
                    HStack(spacing: 2) {
                        Image(systemName: "star.fill")
                            .foregroundColor(.yellow)
                            .font(.caption)

                        Text(String(format: "%.1f", restaurant.rating))
                            .font(.caption)
                            .fontWeight(.semibold)

                        Text("(\(restaurant.reviewCount))")
                            .font(.caption)
                            .foregroundColor(.secondary)
                    }

                    Spacer()

                    // Delivery Time
                    HStack(spacing: 2) {
                        Image(systemName: "clock")
                            .foregroundColor(.secondary)
                            .font(.caption)

                        Text(restaurant.deliveryTime)
                            .font(.caption)
                            .foregroundColor(.secondary)
                    }
                }

                HStack {
                    // Address
                    HStack(spacing: 4) {
                        Image(systemName: "location")
                            .foregroundColor(.secondary)
                            .font(.caption)

                        Text(restaurant.address)
                            .font(.caption)
                            .foregroundColor(.secondary)
                            .lineLimit(1)
                    }

                    Spacer()

                    // Delivery Fee
                    Text("Delivery: $\(restaurant.deliveryFee, specifier: "%.2f")")
                        .font(.caption)
                        .foregroundColor(.green)
                        .fontWeight(.semibold)
                }

                // Action Buttons
                HStack(spacing: 12) {
                    Button("Save") {
                        // Save to favorites
                    }
                    .buttonStyle(.bordered)
                    .controlSize(.small)
                    .frame(maxWidth: .infinity)

                    Button(restaurant.isOpen ? "View Menu" : "Closed") {
                        // View menu
                    }
                    .buttonStyle(.borderedProminent)
                    .controlSize(.small)
                    .frame(maxWidth: .infinity)
                    .disabled(!restaurant.isOpen)
                    .tint(restaurant.cuisineColor)
                }
            }
            .padding(16)
        }
        .background(Color(.systemBackground))
        .clipShape(RoundedRectangle(cornerRadius: 12))
        .shadow(color: .black.opacity(0.1), radius: 4, x: 0, y: 2)
    }
}

struct SortChip: View {
    @Binding var selectedSort: String
    let title: String

    var body: some View {
        Menu {
            Button("Rating") { selectedSort = "Rating" }
            Button("Distance") { selectedSort = "Distance" }
            Button("Price") { selectedSort = "Price" }
            Button("Delivery Time") { selectedSort = "Delivery Time" }
        } label: {
            HStack {
                Image(systemName: "arrow.up.arrow.down")
                    .font(.caption)
                Text(title)
                    .font(.caption)
            }
            .padding(.horizontal, 12)
            .padding(.vertical, 6)
            .background(Color.blue.opacity(0.1))
            .foregroundColor(.blue)
            .clipShape(RoundedRectangle(cornerRadius: 16))
        }
    }
}

struct FilterChip: View {
    let title: String
    let isSelected: Bool
    let action: () -> Void

    var body: some View {
        Button(action: action) {
            Text(title)
                .font(.caption)
                .padding(.horizontal, 12)
                .padding(.vertical, 6)
                .background(isSelected ? Color.blue.opacity(0.2) : Color(.systemGray6))
                .foregroundColor(isSelected ? .blue : .primary)
                .clipShape(RoundedRectangle(cornerRadius: 16))
        }
        .buttonStyle(PlainButtonStyle())
    }
}

struct FilterSheet: View {
    @Environment(\.dismiss) private var dismiss

    var body: some View {
        NavigationStack {
            Form {
                Section("Cuisine Type") {
                    ForEach(["Italian", "Japanese", "Mexican", "French", "Chinese", "Indian"], id: \.self) { cuisine in
                        HStack {
                            Text(cuisine)
                            Spacer()
                            // Toggle would go here
                        }
                    }
                }

                Section("Price Range") {
                    ForEach(PriceRange.allCases, id: \.self) { range in
                        HStack {
                            Text(range.displayText)
                            Spacer()
                            // Toggle would go here
                        }
                    }
                }

                Section("Features") {
                    HStack {
                        Text("Open Now")
                        Spacer()
                        // Toggle would go here
                    }

                    HStack {
                        Text("Free Delivery")
                        Spacer()
                        // Toggle would go here
                    }
                }
            }
            .navigationTitle("Filter Options")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .navigationBarTrailing) {
                    Button("Done") {
                        dismiss()
                    }
                }
            }
        }
    }
}

enum PriceRange: Int, CaseIterable {
    case budget = 1
    case moderate = 2
    case expensive = 3

    var displayText: String {
        switch self {
        case .budget: return "$"
        case .moderate: return "$$"
        case .expensive: return "$$$"
        }
    }

    var color: Color {
        switch self {
        case .budget: return .green
        case .moderate: return .orange
        case .expensive: return .red
        }
    }
}

struct Restaurant: Identifiable {
    let id: String
    let name: String
    let cuisine: String
    let rating: Double
    let reviewCount: Int
    let priceRange: PriceRange
    let deliveryTime: String
    let deliveryFee: Double
    let distance: Double
    let isOpen: Bool
    let isFeatured: Bool
    let address: String

    var cuisineColor: Color {
        switch cuisine {
        case "Italian": return .red
        case "Japanese": return .pink
        case "Mexican": return .orange
        case "French": return .purple
        case "Chinese": return .red
        case "Indian": return .orange
        default: return .blue
        }
    }
}

#Preview {
    RestaurantsListScreen()
}